gshekhawat0820/EchoHub
# Add database persistance

## Description

1. Adds the SwiftData persistence layer.
2. Pre-populates JSON data into the data layer.
3. Adds an action add (and edit) form.
4. Moves horizontal icon list into `CategoryView` for less code repetition in displaying the different categories.
5. Renames several files.
    - The `HouseholdIconView` and `EntertainmentIconView` are the same views for showing action icons, just for different categories. I combined these two files into a single `ActionIconView`.
    - The `HouseholdModel` and `EntertainmentModel` are the same models for parsing the JSON files, just for different categories. I combined these two files into a single `ActionJson` model.
    - Renames `GoogleView` (which contains the add/edit form) as `ActionView`.
6. Modifies the action icon text to navigate to an edit form.

## Testing

1. Add a new action. Ensure that it appears in the Alexa view.
2. Edit an existing action's image, prompt, name, or category. Ensure that the new changes appear in the Alexa view.

## Further steps

There are several UI flaws with the add/edit action form.

1. After clicking on a text input, the keyboard does not lose focus even if the user taps on another part of the screen.
6. There is no user feedback on adding or editing an action. There should be some indication that the form succeeded such as a page navigation or pop-up. There should also be an indication if a form fails. For example, if a user fails to provide an image, the form will not submit. But, there is no indication of such. 
7. There should be a fallback image provided for adding an action.

## Notes

I attempted to use SwiftData's query predicates to query for a specific category, but it resulted in an infinite loop. I don't think I'm misusing it, and there are no noted solutions on forums. For now, we are simply querying all of the actions from the database, and then manually filtering on the UI end. This may actually be preferred since we only have to send one data query, instead of one for each category.
